src/query_plan/query_planning_traversal.rs

```diff
@@ -86,3 +86,173 @@ struct BestQueryPlanInfo {
+                    // Now, this shouldn't happen in practice if composition validation
+                    // has been run successfully (and is not buggy),
+                    // since the goal of composition validation
+                    // is exactly to ensure we can never run into this path.
+                    // In any case, we will throw later if that happens,
+                    // but let's just return the proper result here, which is no plan at all.
+                    None
+                } else {
+                    Some(product * branch.0.len())
+                }
+            })
+            .unwrap_or(0);
+        // debug!("Query has {plan_count} possible plans");
+
+        let max_evaluated_plans = self.parameters.config.debug.max_evaluated_plans as usize;
+        loop {
+            let first_branch = &mut self.closed_branches[0];
+            if plan_count <= max_evaluated_plans || first_branch.0.len() <= 1 {
+                break;
+            }
+            // we remove the right-most option of the first branch,
+            // and them move that branch to it's new place.
+            let prev_size = first_branch.0.len();
+            first_branch.0.pop();
+            plan_count -= plan_count / prev_size;
+            self.reorder_first_branch();
+            // Note that if firstBranch is our only branch, it's fine,
+            // we'll continue to remove options from it (but that is beyond unlikely).
+
+            // debug!("Reduced plans to consider to {plan_count} plans");
+        }
+    }
+
+    /// Moves the first closed branch to after any branch having more options.
+    /// This method assumes that closed branches are sorted by decreasing number of options
+    /// _except_ for the first element which may be out of order,
+    /// and this method restore that order.
+    fn reorder_first_branch(&mut self) {
+        let first_branch_len = self.closed_branches[0].0.len();
+        let i = self
+            .closed_branches
+            .iter()
+            .skip(1)
+            .position(|branch| branch.0.len() <= first_branch_len)
+            .unwrap_or(self.closed_branches.len());
+        // `i` is the smallest index of an element
+        // having the same number or less options than the first one,
+        // so we switch that first branch with the element "before" `i` (which has more elements).
+        self.closed_branches.swap(0, i - 1)
+
+        // FIXME: doesn’t this leave the element foremly at `i - 1` out of place?
+        // Shouldn’t the whole range be shifted?
+        //
+        // It may be easier to do a full sort again, since it’s documented as:
+        //
+        // https://doc.rust-lang.org/std/primitive.slice.html#method.sort_by
+        // > The current algorithm is an adaptive, iterative merge sort inspired by timsort.
+        // > It is designed to be very fast in cases where the slice is nearly sorted
+    }
```

Comment:
It looks like the TypeScript code has a bug (compared to what comment say the intent is), although this only affects the heuristic reducing options semi-arbitrarily.

Reply:
The logic in the JS codebase turns out to be correct, but the JS codebase comments are unfortunately poor/don't describe why.

The key thing the comments don't point out is that the first branch's size only ever decrements by 1. This means that when restoring order, we'll see the first branch (let's say of size `n`), and then potentially a run of elements with size `n+1`, and then any elements after that are size of at most `n`. The `i-1` in the code is the last element of this "run" of `n+1`-sized elements (or the first branch itself, if the run doesn't exist). This means the swap will place an `n+1`-sized element first (which should be fine, since the run of elements after is `n+1`-sized), and then the first branch of size `n` will sit after the `n+1`-sized run but before an element of size `<=n`. So the `Vec` effectively remains sorted.

The gist is it should be fine to remove this `FIXME` comment, but it would help to explain this requirement/behavior in the docstring of `reorder_first_branch()`.

Reply:
I rewrote `reorder_first_branch` with comments to be easier to understand, made it generic in order to unit-test it, and moved the option removal into it to make expectations made of the caller easier to document.

Reply:
@SimonSapin 
This a lot easier to understand, thanks for updating comments and writing tests!